keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+
+This KEP proposes moving ReferenceGrant from its current
+`gateway.networking.k8s.io` API group to a new `grants.authorization.k8s.io` API
+group.
+
+## Motivation
+
+Now that it's clear that ReferenceGrant is useful beyond just Gateway API, it
+would be good to formalize this model in a more neutral home, ideally sig-auth.
+At this point, each project that wants to enable cross-namespace references has
+to choose between introducing a dependency on Gateway API and creating a new
+resource that would largely duplicate ReferenceGrant. Both options would lead to
+confusion for Kubernetes users.
+
+### Goals
+
+* Move ReferenceGrant to sig-auth API Group
+* Clearly define how ReferenceGrant should be used, including both current use
+  cases and guidance for future use cases
+* Implement a library to ensure that ReferenceGrant is implemented consistently
+  by all controllers
+
+### Non-Goals
+
+* Add, remove, or change fields in the ReferenceGrant API
+* Develop an authorizer that will automatically implement ReferenceGrant for
+  all use cases
+
+## Proposal
+
+Move the existing ReferenceGrant resource into a new
+`grants.authorization.k8s.io` API group, defined within the Kubernetes code base
+as part of the 1.27 release.
+
+We will take this opportunity to clarify underspecified parts of the API, but
+will not add, change, or remove any fields as part of this transition.
+This resource will start with v1beta1 as the API version, matching the API version
+it already has within Gateway API.
+
+
+### Risks and Mitigations
+
+#### No Default Implementation
+Similar to the Ingress and Gateway APIs, this API will be dependent on
+implementations by controllers that are not included by default in Kubernetes.
+This could lead to confusion for users. We'll need to rely heavily on
+documentation for this feature, tracking all uses of official Kubernetes APIs
+that support ReferenceGrant in a central place.
+
+#### Potential for Variations Among Implementations
+Because this relies on each individual controller to implement the logic,
+it is possible that implementations may become inconsistent. To avoid that,
+we'll provide a standard library for implementing ReferenceGrant. We'll
+also strongly recommend that every API that relies on ReferenceGrant 
+includes robust conformance tests covering this functionality. Existing
+Gateway API conformance tests can serve as a model for this.
+
+#### Cross-Namespace References may Weaken Namespace Boundaries
+Although we believe that the handshake required for cross-namespace references
+with ReferenceGrant ensures these references will be safe, it does potentially
+weaken existing namespace boundaries. We believe ReferenceGrant will have a net
+benefit on the ecosystem as it will allow workloads, secrets, and configuration
+to be deployed in separate namespaces that more clearly match up with desired
+authorization.
+
+## Design Details
+
+### General Notes
+
+#### `ReferenceGrant` is half of a handshake
```

Comment:
hoisting from the [sig-auth discussion](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.9q15hi5lip46):

does anything ensure the user creating the ReferenceGrant has permissions (read? write?) on the object they are granting access to? Translating the existing ReferenceGrant into an authz check means translating from Kind to Resource

since this is extending "half of a handshake", it seems important to ensure the actor extending the handshake actually has permission to do so

Reply:
This reminds me of a question I forgot: who can create / mutate these grant objects?  Is it part of the `edit` or `admin` cluster roles?

+1 on having a coverage check similar to what RBAC does today.

Reply:
Yes, ReferenceGrant is designed to be revocable, though I think this particular kind of check that the creator/updater of ReferenceGrant indeed has read or write access to the resources it grants access to would only be reasonable at the time the ReferenceGrant was created/updated. If that user had their read/write access revoked, I don't think it is reasonable to revoke the ReferenceGrant previously created by them.

Reply:
I agree that the ideal solution here is that _something_ verifies that the identity creating the ReferenceGrant has access to all the resources it is granting access to.

However, both this and other changes of this kind sound like either a validating webhook or apiserver-native validation. It doesn't seem like it would be a good thing for controllers reconciling these resources to need to go and look up who created them. This is an area I don't know much about though.

I'll leave this as is for now.

Reply:
Added a new section to that states this requirement, for now it just requires read access but can follow up with more specifics on that section.